packages/@lwc/engine-core/src/framework/hydration.ts

```diff
@@ -48,30 +50,53 @@ const enum EnvNodeTypes {
     COMMENT = 8,
 }
 
-function hydrate(vnode: VNode, node: Node) {
+let hasMismatch = false;
+
+export function hydrateRoot(vm: VM) {
+    hasMismatch = false;
+
+    runConnectedCallback(vm);
+    hydrateVM(vm);
+
+    if (hasMismatch) {
+        logError('Hydration completed with errors.', vm);
+    }
+}
+
+function hydrateVM(vm: VM) {
+    if (isTrue(vm.isDirty)) {
+        const children = renderComponent(vm);
+        vm.children = children;
+
+        const parentNode = vm.renderMode === RenderMode.Light ? vm.elm : vm.elm.shadowRoot;
+
+        hydrateChildren(getFirstChild(parentNode), children, parentNode);
+        runRenderedCallback(vm);
+    }
+}
+
+function hydrateNode(node: Node, vnode: VNode): Node | null {
     switch (vnode.type) {
         case VNodeType.Text:
-            hydrateText(vnode, node);
-            break;
+            return hydrateText(node, vnode);
 
         case VNodeType.Comment:
-            hydrateComment(vnode, node);
-            break;
+            return hydrateComment(node, vnode);
 
         case VNodeType.Element:
-            hydrateElement(vnode, node);
-            break;
+            return hydrateElement(node, vnode);
 
         case VNodeType.CustomElement:
-            hydrateCustomElement(vnode, node);
-            break;
+            return hydrateCustomElement(node, vnode);
```

Comment:
This change is reasonable (using `return` instead of `break`), but it's unrelated to this PR.